sycl/doc/Assert.md

```diff
@@ -0,0 +1,296 @@
+| 2 | Print assert message | Assert function should print message to stderr at host | Must have |
+| 3 | Stop under debugger | When debugger is attached, break at assertion point | Highly desired |
+| 4 | Reliability | Assert failure should be reported regardless of kernel deadlock | Highly desired |
+
+Implementations without enough capabilities to implement fourth requirement are
+allowed to realize the fallback approach described below, which does not
+guarantee assertion failure delivery to host, but is still useful in many
+practical cases.
+
+
+## Terms
+
+ - Device-side Runtime - runtime library supplied by the Native Device Compiler
+   and running on the device.
+ - Native Device Compiler - compiler which generates device-native binary image
+   based on input SPIR-V image.
+ - Low-level Runtime - the backend/runtime behind DPCPP Runtime attached via the
+   Plugin Interface.
+ - Accessor metadata - parts of accessor representation at device-side: pointer,
+   ranges, offset.
+
+
+## How it works?
+
+`assert(expr)` macro ends up in call to `__devicelib_assert_fail`. This function
+is part of [Device library extension](extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst#cl_intel_devicelib_cassert).
+
+Format of assert failure message, printed to `stderr` is the following:
+```
+<file>:<line>: <function>: global id: [<gid0>,<gid1>,<gid2>], local id: [<lid0>,<lid1>,<lid2>] Assertion `<expr>` failed.
+```
+
+Implementation of this function is supplied by Native Device Compiler for
+safe approach or by DPCPP Compiler for fallback one.
+
+NB: Due to lack of support of online linking in Level-Zero, the application is
+linked against fallback implementation of `__devicelib_assert_fail`. Hence,
+Native Device Compilers should prefer their implementation instead of the one
+provided in incoming SPIR-V/LLVM IR binary.
+
+Limitations for user after catching the "assert" asynchronous exception:
+ - When using GPU device and the kernel hangs/crashes the subsequent enqueues
+   will fail;
+When using CPU devices the user can proceed with enqueues to the same
+device/queue/context.
+DPCPP Runtime remains in valid state after "assert" exception been thrown.
+
+
+### Safe approach
+
+This is the preferred approach and implementations should use it when possible.
+It guarantees assertion failure notification delivery to the host regardless of
+kernel behavior which hit the assertion.
+
+The Native Device Compiler is responsible for providing implementation of
+`__devicelib_assert_fail` which completely hides details of communication
+between the device code and the Low-Level Runtime from the SYCL device compiler
+and runtime. The Low-Level Runtime is responsible for:
+ - detecting if assert failure took place;
+ - flushing assert message to `stderr` on host.
+
+When detected, Low-level Runtime reports assert failure to DPCPP Runtime
+via events objects. To achieve this, information about assert failure should be
```

Comment:
please elaborate the sequence of what happens and when:
1) assertion hit
2) device immediately sends signal to low-level runtime
3) other gpu threads continue to run
4) low-level runtime immediately receives the signal (in some async thread?) prints the message and updates event status
5) How SYCL runtime detects the event status changed? Separate polling thread which does piEventGetInfo?
If no separate thread, then how - upon subsequent enqueue into exec graph? Does runtime also marks the event as finished immediately after signal is received?
as I noted before, this is the key functional part of the safe approach design, and it should be crystal clear, I believe.

Reply:
Added